### PR TITLE
 Bug in read_csv and read_excel not applying dtype to second col with dup cols

### LINKED ISSUE #35211: BUG: in read_excel for mangled columns only the original/first column dtype is correct, col.N is not parsed correctly
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

This requires setting up an excel file. Reproducible by setting 2 columns with equal headers like `a` and a  row with values `1`. Schematic:

```
a a
1 1
```

Then using:

```python
df = pd.read_excel(file, dtype={'a': str})
assert type(df['a'][0]) == type(df['a.1'][0]) 
```

#### Problem description

The problem is that column `a` is parsed as a `str` and column `a.1` is parsed as a `float64`.

#### Expected Output

Both `a` and `a.1` should be parsed as `str`, such that the assertion is passed.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.4.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.5
numpy            : 1.19.0
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 40.8.0
Cython           : 0.29.20
pytest           : 5.3.5
hypothesis       : None
sphinx           : 3.0.4
blosc            : None
feather          : None
xlsxwriter       : 1.2.9
lxml.etree       : 4.5.1
html5lib         : None
pymysql          : None
psycopg2         : 2.8.5 (dt dec pq3 ext lo64)
jinja2           : 2.11.1
IPython          : 7.15.0
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.1
matplotlib       : 3.2.1
numexpr          : None
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.3.5
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : 1.2.9
numba            : 0.50.0

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #35211
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/_libs/parsers.pyx [hunk 1]
```diff
@@ -685,10 +685,17 @@ cdef class TextReader:
                     count = counts.get(name, 0)
 
                     if not self.has_mi_columns and self.mangle_dupe_cols:
-                        while count > 0:
-                            counts[name] = count + 1
-                            name = f'{name}.{count}'
-                            count = counts.get(name, 0)
+                        if count > 0:
```
Comment (jreback) [inline pandas/_libs/parsers.pyx:688]: would be nice to unify this code between here and the python parser (followon)

**Review by jreback — APPROVED** [review]
thanks @phofl

File: pandas/_libs/parsers.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/parsers.pyx:688]: sounds great!

feels free to open an issue for tracking

Comment (phofl) [reply pandas/_libs/parsers.pyx:688]: Though about using #39345 for this

**Comment by fiendish** [comment]
FYI this breaks parsing with a non-dict dtype.

**Comment by phofl** [comment]
Could you provide an example?

**Comment by fiendish** [comment]
Documentation for read_csv says dtype may be "a type name _or_ dict".
`pandas.read_csv("test.csv", engine="python", dtype="str")` and `pandas.read_csv("test.csv", engine="python", dtype=str)` now error because this code's new calls to .get assume a dict.

**Comment by phofl** [comment]
Your test file has duplicate columns too? Will look into this later

**Comment by fiendish** [comment]
Here's my test file:
[test.csv](https://github.com/pandas-dev/pandas/files/6657781/test.csv)

**Comment by fiendish** [comment]
I should have filed an issue first. I've done that now: https://github.com/pandas-dev/pandas/issues/42022

**Pull request merged**